### PR TITLE
chore: Update react virtual to v3

### PR DESCRIPTION
**Pull request by pan-kot**
### Description

Upgraded react-virtual to the latest version (v3, see: https://tanstack.com/virtual/v3/docs/adapters/react-virtual) to fix npm install issue caused by React 18 not being listed in peer dependencies of v2.

Additionally, removed the workaround introduced for v2 to avoid "Maximum update depth exceeded" error. A similar problem is possible in v3 but it does not cause the error. Instead, the virtual list bounces between two or more sizes with a small delay which is not expected to cause major UX issues. The problem is rare and can be mitigated by updating item styles so that list total size to no longer affect individual item heights.

Rel: AWSUI-26338, https://github.com/cloudscape-design/components/pull/1202

### How has this been tested?

Existing unit and integration tests, manual tests, screenshot tests.

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- _Changes include appropriate documentation updates._
- _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#public-apis)._
- _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#browsers-support)._
- _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- _Changes are covered with new/existing unit tests?_
- _Changes are covered with new/existing integration tests?_
</details>

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: __mocks__/@tanstack/react-virtual.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (just-boris) [inline __mocks__/@tanstack/react-virtual.ts:1]: If we use a global mock, how do we ensure that unit tests of our consumers will not fail? Can we not use such ultimate tool?

File: src/autosuggest/virtual-list.tsx [hunk 2]
```diff
@@ -26,22 +26,25 @@ const VirtualList = ({
...
     estimateSize: useCallback(() => 31, [width, highlightText]),
+    overscan: 5,
   });
 
+  const totalItems = autosuggestItemsState.items.length;
   useEffect(() => {
-    if (autosuggestItemsState.highlightType.moveFocus) {
-      rowVirtualizer.scrollToIndex(autosuggestItemsState.highlightedIndex);
+    const index = autosuggestItemsState.highlightedIndex;
+    if (autosuggestItemsState.highlightType.moveFocus && index >= 0 && index < totalItems) {
```
Comment (just-boris) [inline src/autosuggest/virtual-list.tsx:44]: What changed here? Why do we need to validate `index` now?

**Review by just-boris — COMMENTED** [review]
Also questions about testing

1. PR build fails
2. Did you measure performance differences?
3. For the manual testing, how much are you confident in your own testing? Did you ask other team members to help?

File: package.json [hunk 3]
```diff
@@ -29,6 +29,7 @@
     "@dnd-kit/sortable": "^7.0.2",
     "@dnd-kit/utilities": "^3.2.1",
     "@juggle/resize-observer": "^3.3.1",
+    "@tanstack/react-virtual": "^3.0.1",
```
Comment (just-boris) [inline package.json:32]: Should we pin this dependency to avoid unexpected versions installed by our consumers?

**Comment by pan-kot** [comment]
The measure bundle size job fails because we replace its version of components with a local copy but do not account for newly added dependencies:

```
run: |
  cd components
  npm i --force
  npm run quick-build

  cd ../bundle-size
  rm -rf node_modules/@cloudscape-design/components
  cp -r $GITHUB_WORKSPACE/components/lib/components node_modules/@cloudscape-design/components
  cp main-with-cloudscape.jsx main.jsx
  npm run build
```

File: package.json (hunk 3 above)
Comment (pan-kot) [reply package.json:32]: We don't pin other dependencies, why should we pin this one?

File: src/autosuggest/virtual-list.tsx (hunk 2 above)
Comment (pan-kot) [reply src/autosuggest/virtual-list.tsx:44]: In v3 there is an error thrown when executing scroll to an index for which there is no item defined, see https://github.com/TanStack/virtual/blob/main/packages/virtual-core/src/index.ts#L822

File: package.json (hunk 3 above)
Comment (just-boris) [reply package.json:32]: This used to be pinned before because of bugs. Check the pre-github commit history

**Comment by just-boris** [comment]
> do not account for newly added dependencies

Can you make it work?

File: __mocks__/@tanstack/react-virtual.ts (hunk 1 above)
Comment (pan-kot) [reply __mocks__/@tanstack/react-virtual.ts:1]: Updated

**Comment by pan-kot** [comment]
> Also questions about testing
> 
> 1. PR build fails
> 2. Did you measure performance differences?
> 3. For the manual testing, how much are you confident in your own testing? Did you ask other team members to help?

1. Fixed
2. I did not measure it but tested with throttling on a list of 1000 items. The v2 feels slightly faster when scrolling but nothing significant. I did not observe any difference to the initial rendering performance
3. I tested using the available test pages plus specifically tested for https://github.com/cloudscape-design/components/pull/1202 scenario. I think the manual testing done by both of us plus CI and dry-run should be enough, but I can ask for more testing if you think it is necessary

File: __mocks__/@tanstack/react-virtual.ts (hunk 1 above)
Comment (just-boris) [reply __mocks__/@tanstack/react-virtual.ts:1]: What is updated here? This mock still applies to all of our tests. We never test against the actual virtual instance in JSDOM.

Do you think it is okay? What if something fails in our live consumers?

File: package.json (hunk 3 above)
Comment (just-boris) [reply package.json:32]: What is your final decision here? How about asking the team opinion?

File: src/autosuggest/virtual-list.tsx (hunk 2 above)
Comment (just-boris) [reply src/autosuggest/virtual-list.tsx:44]: ok for now, but doesn't it sounds like a problem with our code?

Comment (pan-kot) [reply src/autosuggest/virtual-list.tsx:44]: I agree, there is an opportunity for improvement. In this particular case the issue comes from the fact we use -1 as the initial value of the highlighted index.

File: package.json (hunk 3 above)
Comment (pan-kot) [reply package.json:32]: Agreed on having this dependency pinned.

File: __mocks__/@tanstack/react-virtual.ts (hunk 1 above)
Comment (just-boris) [reply __mocks__/@tanstack/react-virtual.ts]: How do we release this to live, if it has required upgrade steps to our consumers?

**Comment by pan-kot** [comment]
Closed in favour of https://github.com/cloudscape-design/components/pull/1897

**Pull request closed without merging**